### PR TITLE
Pvp options

### PR DESCRIPTION
**Pull request by Laserlicht**
Implementation like in HD mod except:
- Town area is scrollable (no town limit)
- no highlight of selected town (no green icons in OH3)
- No gold transfer yet (not part of this PR)

Also adds possibility to translate strings within messages

https://github.com/vcmi/vcmi/assets/13953785/a3c78365-4a69-4990-b3c8-8cb8d6c76bd6

### CODE REVIEW

**Comment by dydzio0614** [comment]
Since I think access to this panel should not block chat window etc... probably optimal place for it would be here at the moment
obraz

**Comment by dydzio0614** [comment]
Also if you do not want to deal with this feature then just skip it :P

**Comment by Laserlicht** [comment]
> Since I think access to this panel should not block chat window etc...

Yes, but I also think that we should bot flood this ui with multiplayer stuff. It's also used by peoples which never play multiplayer. Because of this it would be great to limit it to one button

**Comment by dydzio0614** [comment]
My initial idea was to disable feature completely unless game is multiplayer / hot seat and number of players on map is more than 4

We can try to go the way you show but i am almost sure the first time feature gets used in pvp it will get github report due to "not being as convenient to use as in hota"

**Comment by Laserlicht** [comment]
@dydzio0614 What about this space. Do we need to display this, when chat is opened?

grafik

**Comment by dydzio0614** [comment]
it doesn't serve much purpose so as long as difficulty was already picked it could be used for something else

Hard for me to judge final convenience of final solution you have in mind :P

**Comment by Laserlicht** [comment]
@dydzio0614 :D

grafik

And without chat as OH3.

**Comment by dydzio0614** [comment]
hmm but how are we going to fit 12+ towns this way :D

**Comment by dydzio0614** [comment]
VCMI multiplayer tournament will have 13 towns... for a start :P

**Comment by Laserlicht** [comment]
> hmm but how are we going to fit 12+ towns this way :D

Scrollbar for town list?

**Comment by dydzio0614** [comment]
oh ok :P

**Comment by Laserlicht** [comment]
Do you think something like this would be okay?

**Comment by dydzio0614** [comment]
sounds good enough for me, in worst case we will have some logic of the panel itself :D

**Comment by Laserlicht** [comment]
@dydzio0614 See first post

**Comment by dydzio0614** [comment]
looks good

**Comment by Laserlicht** [comment]
Ready now

**Review by dydzio0614 — APPROVED** [review]

File: server/processors/PlayerMessageProcessor.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ void PlayerMessageProcessor::playerMessage(PlayerColor player, const std::string
 	if (handleCheatCode(message, player, currObj))
 	{
 		if(!gameHandler->getPlayerSettings(player)->isControlledByAI())
-			broadcastSystemMessage(VLC->generaltexth->allTexts[260]);
+			broadcastSystemMessage("~~core.genrltxt.260~~");
```
Comment (IvanSavenko) [inline server/processors/PlayerMessageProcessor.cpp:45]: If you need client-side translations you should send MetaString to client instead

File: client/lobby/CSelectionBase.cpp [hunk 2]
```diff
@@ -373,6 +393,120 @@ void CChatBox::addNewMessage(const std::string & text)
...
+void TownSelector::updateListItems()
+{
+	OBJ_CONSTRUCTION;
+	int line = slider ? slider->getValue() : 0;
+	int x_offset = slider ? 0 : 8;
+	
+	towns.clear();
+	townsArea.clear();
+
+	int x = 0, y = 0;
```
Comment (IvanSavenko) [inline client/lobby/CSelectionBase.cpp:475]:
```suggestion
	int x = 0;
	int y = 0;
```

File: client/lobby/CSelectionBase.cpp [hunk 3]
```diff
@@ -373,6 +393,120 @@ void CChatBox::addNewMessage(const std::string & text)
...
+}
+
+TownSelector::TownSelector(const Point & loc)
+{
+	OBJ_CONSTRUCTION;
+	pos += loc;
+	setRedrawParent(true);
+
+	int count = 0;
+	CGI->factions()->forEach([this, &count](const Faction *entity, bool &stop){
```
Comment (IvanSavenko) [inline client/lobby/CSelectionBase.cpp:447]: You can use `getDefaultAllowed()` and use standard for loop over returned vector. Tends to be more clear and easy to debug than methods with lambdas

File: client/GameChatHandler.cpp [hunk 4]
```diff
@@ -86,22 +90,43 @@ void GameChatHandler::onNewGameMessageReceived(PlayerColor sender, const std::st
...
+	while(std::regex_search(searchStart, txt.cend(), match, expr) )
+	{
+		result += match.prefix();
+		result += VLC->generaltexth->translate(match.str(1));
+		
+		searchStart = match.suffix().first;
+		tmp_suffix = match.suffix();
+	}
+	return result.empty() ? txt : result + tmp_suffix;
+}
```
Comment (IvanSavenko) [inline client/GameChatHandler.cpp:132]: Replace with MetaString

File: server/NetPacksLobbyServer.cpp [hunk 5]
```diff
@@ -366,3 +369,42 @@ void ApplyOnServerNetPackVisitor::visitLobbyForceSetPlayer(LobbyForceSetPlayer &
...
+{
+	std::vector<FactionID> allowedTowns;
+
+	VLC->townh->forEach([pack, &allowedTowns](const Faction *entity, bool &stop){
+		if(!entity->hasTown())
+			return;
+		if(std::find(pack.bannedTowns.begin(), pack.bannedTowns.end(), entity->getId()) == pack.bannedTowns.end()) {
+			allowedTowns.push_back(entity->getId());
+		}
+	});
```
Comment (IvanSavenko) [inline server/NetPacksLobbyServer.cpp:389]:
```suggestion
	for (auto const & factionID : VLC->townh->getDefaultAllowed())
		if(std::find(pack.bannedTowns.begin(), pack.bannedTowns.end(), factionID) == pack.bannedTowns.end()) {
			allowedTowns.push_back(factionID);
		}
	});
```

File: server/processors/PlayerMessageProcessor.cpp (hunk 1 above)
Comment (Laserlicht) [reply server/processors/PlayerMessageProcessor.cpp:45]: And change the messaging logic?

Comment (IvanSavenko) [reply server/processors/PlayerMessageProcessor.cpp:45]: Why not?
Existing messages with player input can be added as raw strings to MetaString.
Better than these `~~` hacks

File: client/GameChatHandler.cpp (hunk 4 above)
Comment (Laserlicht) [reply client/GameChatHandler.cpp]: Done :)

**Pull request merged**